src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,444 @@
+                    int preWaitTimeMs = Environment.TickCount;
+
+                    // Recalculate Timeout
+                    int timeoutDurationMs = Timeout.Infinite;
+                    if (numUserWaits == 0)
+                    {
+                        timeoutDurationMs = ThreadPoolThreadTimeoutMs;
+                    }
+                    else
+                    {
+                        for (int i = 0; i < numUserWaits; i++)
+                        {
+                            if (_registeredWaits[i].IsInfiniteTimeout)
+                            {
+                                continue;
+                            }
+
+                            int handleTimeoutDurationMs = _registeredWaits[i].TimeoutTimeMs - preWaitTimeMs;
+
+                            if (timeoutDurationMs == Timeout.Infinite)
+                            {
+                                timeoutDurationMs = handleTimeoutDurationMs > 0 ? handleTimeoutDurationMs : 0;
+                            }
+                            else
+                            {
+                                timeoutDurationMs = Math.Min(handleTimeoutDurationMs > 0 ? handleTimeoutDurationMs : 0, timeoutDurationMs);
+                            }
+
+                            if (timeoutDurationMs == 0)
+                            {
+                                break;
+                            }
+                        }
+                    }
+
+                    int signaledHandleIndex = WaitHandle.WaitAny(_waitHandles, numUserWaits + 1, timeoutDurationMs);
+
+                    if (signaledHandleIndex == 0) // If we were woken up for a change in our handles, continue.
+                    {
+                        continue;
+                    }
+
+                    RegisteredWaitHandle signaledHandle = signaledHandleIndex != WaitHandle.WaitTimeout ? _registeredWaits[signaledHandleIndex - 1] : null;
+
+                    if (signaledHandle != null)
+                    {
+                        QueueWaitCompletion(signaledHandle, false);
+                    }
+                    else
+                    {
+                        if(numUserWaits == 0)
+                        {
+                            if (ThreadPoolInstance.TryRemoveWaitThread(this))
+                            {
+                                return;
+                            }
+                        }
+
+                        int elapsedTicks = Environment.TickCount - preWaitTimeMs; // Calculate using relative time to ensure we don't have issues with overflow wraparound
+                        for (int i = 0; i < numUserWaits; i++)
+                        {
+                            RegisteredWaitHandle registeredHandle = _registeredWaits[i];
+                            int handleTimeoutDurationMs = registeredHandle.TimeoutTimeMs - preWaitTimeMs;
+                            if (elapsedTicks > handleTimeoutDurationMs)
```

Comment:
Should be >=. It's possible that elapsedTicks is 0 and handleTimeoutDurationMs is 0 (that is, we did a WaitAny with a timeout of 0 and Environment.TickCount above returned the same value as preWaitTimeMs), then the callback should be called saying it timed out. For nonzero values I don't think >= makes much difference from > so it should be fine to always use >=.